pkg/security/admission/admission.go

```diff
@@ -0,0 +1,385 @@
+//     with the validated SCC.  If we don't find any reject the pod and give all errors from the
+//     failed attempts.
+func (c *constraint) Admit(a kadmission.Attributes) error {
+	if a.GetResource() != string(kapi.ResourcePods) {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*kapi.Pod)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	// get all constraints that are usable by the user
+	glog.V(4).Infof("getting security context constraints for pod %s (generate: %s) in namespace %s with user info %v", pod.Name, pod.GenerateName, a.GetNamespace(), a.GetUserInfo())
+	matchedConstraints, err := getMatchingSecurityContextConstraints(c.store, a.GetUserInfo())
+	if err != nil {
+		return kadmission.NewForbidden(a, err)
+	}
+
+	// get all constraints that are usable by the SA
+	saConstraints, err := c.getSecurityContextConstraintsForServiceAccount(a.GetNamespace(), pod)
+	if err != nil {
+		glog.Errorf("error getting security context constraints for service account %s: %v.  Only user constraints will be used.", pod.Spec.ServiceAccount, err)
+	} else {
+		matchedConstraints = append(matchedConstraints, saConstraints...)
+	}
+
+	// remove duplicate constraints and sort
+	matchedConstraints = deDupeSecurityContextConstraints(matchedConstraints)
+	sort.Sort(ByRestrictions(matchedConstraints))
+	providers, err := c.createProvidersFromConstraints(a.GetNamespace(), matchedConstraints)
+	if err != nil {
+		return kadmission.NewForbidden(a, err)
+	}
+	logProviders(pod, providers)
+
+	// all containers in a single pod must validate under a single provider or we will reject the request
+	validationErrs := fielderrors.ValidationErrorList{}
+	for _, provider := range providers {
+		errs := generateAndValidateSecurityContext(provider, pod)
+		if len(errs) == 0 {
+			// the entire pod validated, annotate and accept the pod
+			if pod.ObjectMeta.Annotations == nil {
+				pod.ObjectMeta.Annotations = map[string]string{}
+			}
+			glog.V(4).Infof("pod %s (generate: %s) validated against provider %s", pod.Name, pod.GenerateName, provider.GetSCCName())
+			pod.ObjectMeta.Annotations[allocator.ValidatedSCCAnnotation] = provider.GetSCCName()
+			return nil
+		}
+		prefix := fmt.Sprintf("provider %s produced errors:", provider.GetSCCName())
+		validationErrs = append(validationErrs, errs.Prefix(prefix)...)
+	}
+	// we didn't validate against any security context constraint provider, reject the pod and give the
+	// errors for each attempt
+	glog.V(4).Infof("unable to validate pod %s (generate: %s) against any security context constraint: %v", pod.Name, pod.GenerateName, validationErrs)
+	return kadmission.NewForbidden(a, fmt.Errorf("unable to validate against any security context constraint: %v", validationErrs))
+}
+
+func generateAndValidateSecurityContext(provider scc.SecurityContextConstraintsProvider, pod *kapi.Pod) fielderrors.ValidationErrorList {
```

Comment:
Maybe include 'assign' in the name of this method somehow, since it actually modifies the containers.